### ISSUE TITLE
Modifing child model's attr_accessor triggers child model's validation at v5.2.3, but not v5.2.4

### ISSUE THREAD

**Issue by megane42** [rails/rails#38157]
There is some breaking change for validations between v5.2.3 and v5.2.4.
When;

- there are 2 models (e.g. Foo and Bar),
- and Foo `has_one` Bar,
- and Foo has `accepts_nested_attributes_for :bar`,
- and Bar has an `attr_accessor` (e.g. name),
- and `foo.bar.name` is modified,
- and `foo.save` is called,

then the validation for `bar` is executed at v5.2.3, but not at v5.2.4.

I have no idea which behavior is ideal, but at least it seems this is a breaking change on patch version update.

### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "5.2.4"
  gem "sqlite3", "~> 1.3.6"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :foos, force: true do |t|
  end

  create_table :bars, force: true do |t|
    t.integer :foo_id
  end
end

$count_validation = 0

class Foo < ActiveRecord::Base
  has_one :bar
  accepts_nested_attributes_for :bar
end

class Bar < ActiveRecord::Base
  belongs_to :foo
  attr_accessor :name
  before_validation -> { $count_validation += 1 }
end

class BugTest < Minitest::Test
  def test_association_stuff
    foo = Foo.create
    bar = foo.create_bar(name: "old_name")

    $count_validation = 0

    foo.bar.name = "new_name"
    foo.save
    assert_equal 1, $count_validation
  end
end
```

### Expected behavior

* The behavior is consistent between v5.2.3 and v5.2.4
  * i.e. The validation for Bar is executed

### Actual behavior

* The behavior is *not* consistent between v5.2.3 and v5.2.4
    * i.e. The validation for Bar is *not* executed

### System configuration
**Rails version**:

* v5.2.3 and v5.2.4

**Ruby version**:

* ruby 2.6.5p114 (2019-10-01 revision 67812) [x86_64-darwin17]

**Comment by Natch** [comment]
![LGTM](https://lgtmoon.herokuapp.com/images/70642)

**Comment by rafaelfranca** [comment]
Thank you for the issue. Is this also a problem on Rails 6.0? Rails 5.2 is not supported to bug fixes anymore.

**Comment by megane42** [comment]
Yes, I have confirmed it can be reproduced on Rails 6.0.0, 6.0.1, and 6.0.2.1.

**Comment by marceloperini** [comment]
@megane42, @rafaelfranca  I think this is the valid behavior, because there are no validation to call. If you change `Bar` model, and add one validation, your test pass.

``` ruby
class Bar < ActiveRecord::Base
  belongs_to :foo

  attr_accessor :name

  validates :name, presence: true

  before_validation -> { $count_validation += 1 }
end
```

**Comment by tyler717** [comment]
Using the test case provided it appears that the change in behavior was introduced in [this](https://github.com/rails/rails/commit/1cdf6fc9cfc879a3ae0a433e86284a171987b017) commit.  You might try going from ```attr_accessor :name``` to ```attribute :name```.  In running the test against 5.2.4 and 6.0.2 the before_validation would fire after making that change.

**Comment by marcelolx** [comment]
@zzak Based on @tyler717 @marceloperini comments I think that this isn't a bug and could be closed. I understand that it was a behavior change in a patch version update, but since Rails 5.2 is not supported for bugfixes anymore I don't see something that needs to be fixed here.
